internal/metrics/reporter.go

```diff
@@ -0,0 +1,326 @@
+			Seconds: int64(DefaultReporterMeasurementDuration.Seconds()) * intervals,
+		},
+	}
+}
+
+// Measure takes a snapshot of system metrics and sends them
+// to a LightStep endpoint.
+func (r *Reporter) Measure(ctx context.Context, intervals int64) error {
+	start := time.Now()
+	r.Start = start
+	ctx, cancel := context.WithTimeout(ctx, r.timeout)
+	defer cancel()
+
+	m, err := Measure(ctx, 0*time.Second)
+	if err != nil {
+		return err
+	}
+
+	pb, err := r.prepareRequest(m)
+	if err != nil {
+		return err
+	}
+
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.cpu.user", m.ProcessCPU.User-r.stored.ProcessCPU.User, metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.cpu.sys", m.ProcessCPU.System-r.stored.ProcessCPU.System, metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.gc.count", float64(m.GarbageCollector.NumGC-r.stored.GarbageCollector.NumGC), metricspb.MetricKind_COUNTER, intervals))
+
+	pb.Points = append(pb.Points, r.addFloat("mem.available", float64(m.Memory.Available), metricspb.MetricKind_GAUGE, intervals))
+	pb.Points = append(pb.Points, r.addFloat("mem.total", float64(m.Memory.Used), metricspb.MetricKind_GAUGE, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.mem.heap_alloc", float64(m.Memory.HeapAlloc), metricspb.MetricKind_GAUGE, intervals))
+
+	for label, cpu := range m.CPU {
+		pb.Points = append(pb.Points, r.addFloat("cpu.sys", cpu.System-r.stored.CPU[label].System, metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("cpu.user", cpu.User-r.stored.CPU[label].User, metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("cpu.total", cpu.Total-r.stored.CPU[label].Total, metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("cpu.usage", cpu.Usage-r.stored.CPU[label].Usage, metricspb.MetricKind_COUNTER, intervals))
+	}
+	for label, nic := range m.NIC {
+		pb.Points = append(pb.Points, r.addFloat("net.bytes_recv", float64(nic.BytesReceived-r.stored.NIC[label].BytesReceived), metricspb.MetricKind_COUNTER, intervals))
+		pb.Points = append(pb.Points, r.addFloat("net.bytes_sent", float64(nic.BytesSent-r.stored.NIC[label].BytesSent), metricspb.MetricKind_COUNTER, intervals))
+	}
+
+	err = r.send(ctx, pb)
+	if err != nil {
+		return err
+	}
+
+	r.stored = m
+	r.MetricsCount = len(pb.Points)
+	r.End = time.Now()
+	return nil
+}
+
+func (r *Reporter) send(ctx context.Context, ingestRequest *metricspb.IngestRequest) error {
+	b, err := proto.Marshal(ingestRequest)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(http.MethodPost, r.address, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	req = req.WithContext(ctx)
+
+	req.Header.Set(contentTypeHeader, protoContentType)
+	req.Header.Set(acceptHeader, protoContentType)
+	req.Header.Set(accessTokenHeader, r.accessToken)
+
+	if !r.skippedInitialReport {
```

Comment:
Does this mean that no metrics will get sent if the tracer is closed after either <30s or <1m?